File: verify.go
package securedb

import (
	"database/sql"
	"fmt"
	"os"
)

// sqliteHeaderMagic is the first 16 bytes of an unencrypted SQLite database file.
const sqliteHeaderMagic = "SQLite format 3\x00"

// LooksPlaintext checks if a file starts with the SQLite plaintext header.
// Returns true if the file is an unencrypted SQLite database.
// Returns false if the file does not exist or does not match.
func LooksPlaintext(path string) (bool, error) {
	f, err := os.Open(path)
	if err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, fmt.Errorf("securedb: open file: %w", err)
	}
	defer f.Close()

	header := make([]byte, 16)
	n, err := f.Read(header)
	if err != nil || n < 16 {
		return false, nil
	}

	return string(header) == sqliteHeaderMagic, nil
}

// CanOpenWithKey attempts to open the database with the given key,
// verifies it can read, then closes. Returns a typed error on failure.
//
// Possible errors:
//   - ErrFileNotFound: file does not exist
//   - ErrWrongKey: key cannot decrypt the database
//   - ErrNotDatabase: file is not a valid SQLite/SQLCipher database
//   - ErrCipherUnavailable: SQLCipher not available
func CanOpenWithKey(path string, key []byte) error {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return ErrFileNotFound
	}

	noCreate := false
	db, err := Open(path, Options{
		Key:             key,
		Encryption:      EncryptionRequired,
		CreateIfMissing: &noCreate,
		ReadOnly:        true,
	})
	if err != nil {
		return err
	}
	defer db.Close()

	// Double-check: try to read.
	var count int
	if err := db.QueryRow("SELECT count(*) FROM sqlite_master").Scan(&count); err != nil {
		return fmt.Errorf("%w: verification query failed", ErrWrongKey)
	}

	return nil
}

// VerifyCipherMetadata checks that PRAGMA cipher_version returns a valid
// version string on an open database. Returns the version on success.
func VerifyCipherMetadata(db *sql.DB) (string, error) {
	var version string
	err := db.QueryRow("PRAGMA cipher_version").Scan(&version)
	if err != nil {
		return "", fmt.Errorf("%w: %v", ErrCipherUnavailable, err)
	}
	if version == "" {
		return "", ErrCipherUnavailable
	}
	return version, nil
}

File: src/verify.rs
use std::fs::File;
use std::io::Read;
use std::path::Path;

use crate::Database;
use crate::connection::wrap_key;
use crate::error::{Error, Result};
use crate::options::{EncryptionMode, Options};

/// The 16-byte header of an unencrypted SQLite database.
const SQLITE_HEADER: &[u8; 16] = b"SQLite format 3\0";

/// Checks whether a file starts with the plaintext SQLite header.
///
/// Returns `Ok(false)` if the file does not exist, cannot be read, or is
/// shorter than 16 bytes. Returns `Ok(true)` if the header matches
/// (indicating an unencrypted database).
pub fn looks_plaintext(path: &str) -> Result<bool> {
    let p = Path::new(path);
    if !p.exists() {
        return Ok(false);
    }

    let mut f = match File::open(p) {
        Ok(f) => f,
        Err(_) => return Ok(false),
    };

    let mut buf = [0u8; 16];
    match f.read_exact(&mut buf) {
        Ok(()) => Ok(buf == *SQLITE_HEADER),
        Err(_) => Ok(false),
    }
}

/// Attempts to open the database at `path` with the given key in read-only
/// mode and verifies that it can be read.
///
/// The verification is performed by `open()` itself (which executes
/// `SELECT count(*) FROM sqlite_master` atomically after keying).
///
/// Returns typed errors:
/// - `Error::FileNotFound` if the file does not exist
/// - `Error::WrongKey` if the key cannot decrypt the database
/// - `Error::NotDatabase` if the file is not a valid database
/// - `Error::CipherUnavailable` if SQLCipher is not available
pub fn can_open_with_key(path: &str, key: &[u8]) -> Result<()> {
    let p = Path::new(path);
    if !p.exists() {
        return Err(Error::FileNotFound);
    }

    let db = crate::open(
        path,
        Options {
            key: Some(wrap_key(key.to_vec())),
            encryption: EncryptionMode::Required,
            read_only: true,
            create_if_missing: Some(false),
            ..Options::default()
        },
    )?;

    // open() already verified sqlite_master — just close.
    let _ = db.close();
    Ok(())
}

/// Queries `PRAGMA cipher_version` on an open database.
///
/// Returns the version string on success, or `Error::CipherUnavailable`
/// if the cipher is not active. This is a diagnostic function — the real
/// proof of a valid key is the `SELECT count(*) FROM sqlite_master` probe
/// that `open()` performs atomically.
pub fn verify_cipher_metadata(db: &Database) -> Result<String> {
    let version: String =
        db.with_connection(|conn| conn.query_row("PRAGMA cipher_version;", [], |row| row.get(0)))?;

    if version.is_empty() {
        return Err(Error::CipherUnavailable);
    }

    Ok(version)
}
